src/systemctl/systemctl.c

```diff
@@ -7495,7 +7500,8 @@ static int halt_now(enum action a) {
         /* The kernel will automaticall flush ATA disks and suchlike
          * on reboot(), but the file systems need to be synce'd
          * explicitly in advance. */
-        (void) sync();
+        if (!strstr(program_invocation_short_name, "halt") && !arg_no_sync)
```

Comment:
Why do you need the first part of the condition? `arg_no_sync` can only be set in `halt_parse_argv`, which will only be called in the command is `halt`.